app/models/project_services/composer_service.rb

```diff
@@ -0,0 +1,527 @@
+          }
+        ]
+      }
+    ]
+  end
+
+  # disable test button
+  def can_test?
+    false
+  end
+
+  def process_packages_on_save
+
+    # process packages for all tags
+    project.repository.tags.each do |tag|
+      process_commit(tag)
+    end
+
+    # process packages for all branches
+    project.repository.branches.each do |branch|
+      process_commit(branch)
+    end
+  end
+
+  def process_commit(ref)
+    previous_package_removed = false
+    begin
+      if commit_was_activated? && commit_was_exported?(ref)
+
+        if package_mode_was == 'advanced'
+          defaults = ActiveSupport::JSON.decode(custom_json_was)
+        else
+          defaults = { 'type'=>package_type_was }
+        end
+
+        package = Composer::Package.
+                    new(project, ref, package_mode_was, defaults)
+
+        manager.rm_package(package)
+        previous_package_removed = true
+
+      end
+    rescue
+      # skip on error
+    end
+
+    begin
+      if package_mode == 'advanced'
+        defaults = ActiveSupport::JSON.decode(custom_json)
+      else
+        defaults = { 'type'=>package_type }
+      end
+
+      package = Composer::Package.
+                  new(project, ref, package_mode, defaults)
+
+      if activated? && commit_is_exported?(ref)
+        manager.add_package(package)
+      elsif not previous_package_removed
+        manager.rm_package(package)
+      end
+
+    rescue
+      # skip on error
+    end
+  end
+
+  def execute(push_data)
+    newrev = push_data[:after]
+    ref = push_data[:ref]
+
+    # sync our changes
+    if newrev == Gitlab::Git::BLANK_SHA # push delete
+
+      # recreate exported packages since we dont have access
+      # to the original push
+      manager.clear_packages
+
+      # process packages for all tags
+      project.repository.tags.each do |t|
+        process_commit(t)
+      end
+
+      # process packages for all branches
+      project.repository.branches.each do |b|
+        process_commit(b)
+      end
+
+    else # push create / modify
+      if push_to_branch?(ref)
+        match = project.repository.branches.detect do |b|
+          b.name == branch_name(ref) && b.target == newrev
+        end
+      elsif push_to_tag?(ref)
+        match = project.repository.tags.detect do |t|
+          t.name == tag_name(ref) && t.target == newrev
+        end
+      end
+      if match
+        process_commit(match)
+      end
+    end
+
+  rescue
+    # skip on error
+  end
+
+  private
+
+  def manager
+    @manager ||= Composer::Manager.new(project)
+  end
+
+  def commit_is_exported?(ref)
+    if ref.instance_of?(Gitlab::Git::Branch)
+      branch_is_exported?(ref)
+    elsif ref.instance_of?(Gitlab::Git::Tag)
+      tag_is_exported?(ref)
+    else
+      false
+    end
+  end
+
+  def branch_is_exported?(branch)
+    if branche_filters
+      filters = (branch_filters.strip! || branch_filters).
+                  gsub(" ", "").split(',')
```

Comment:
Prefer single-quoted strings when you don't need string interpolation or special symbols.